server/src/main/java/org/elasticsearch/index/mapper/IgnoredSourceFieldMapper.java

```diff
@@ -145,6 +150,56 @@ static NameValue decode(Object field) {
         return new NameValue(name, parentOffset, value, null);
     }
 
+    public record MappedNameValue(NameValue nameValue, XContentType type, Map<String, Object> map) {}
+
+    /**
+     * Parses the passed byte array as a NameValue and coverts its decoded value to a map of maps that corresponds to the field-value
+     * subtree. There is only a single pair at the top level, with the key corresponding to the field name. If the field contains a single
+     * value, the map contains a single key-value pair. Otherwise, the value of the first pair will be another map etc.
+     * @param value encoded NameValue
+     * @return MappedNameValue with the parsed NameValue, the XContentType to use for serializing its contents and the field-value map.
+     * @throws IOException
+     */
+    public static MappedNameValue decodeAsMap(byte[] value) throws IOException {
+        BytesRef bytes = new BytesRef(value);
+        IgnoredSourceFieldMapper.NameValue nameValue = IgnoredSourceFieldMapper.decode(bytes);
+        XContentBuilder xContentBuilder = XContentBuilder.builder(XContentDataHelper.getXContentType(nameValue.value()).xContent());
+        xContentBuilder.startObject().field(nameValue.name());
+        XContentDataHelper.decodeAndWrite(xContentBuilder, nameValue.value());
+        xContentBuilder.endObject();
+        Tuple<XContentType, Map<String, Object>> result = XContentHelper.convertToMap(BytesReference.bytes(xContentBuilder), true);
+        return new MappedNameValue(nameValue, result.v1(), result.v2());
+    }
+
+    /**
+     * Clones the passed NameValue, using the passed map to produce its value.
+     * @param mappedNameValue containing the NameValue to clone
+     * @param map containing a simple field-value pair, or a deeper field-value subtree for objects and arrays with fields
+     * @return a byte array containing the encoding form of the cloned NameValue
+     * @throws IOException
+     */
+    public static byte[] encodeFromMap(MappedNameValue mappedNameValue, Map<String, Object> map) throws IOException {
+        // The first entry is the field name, we skip to get to the value to encode.
+        assert map.size() == 1;
+        Object content = map.values().iterator().next();
+
+        // Check if the field contains a single value or an object.
+        @SuppressWarnings("unchecked")
+        XContentBuilder xContentBuilder = (content instanceof Map<?, ?> objectMap)
```

Comment:
nit: can't you check for `Map<String, ?>` and remove `unchecked` and cast?

Reply:
Getting:

```
 error: Object cannot be safely cast to Map<String,?>
        XContentBuilder xContentBuilder = (content instanceof Map<String, ?> objectMap)
```

I miss C++ templates..

Reply:
Never expected to see that statement from anybody.